stripe/src/main/java/com/stripe/android/model/SetupIntent.java

```diff
@@ -0,0 +1,376 @@
+                .setPaymentMethodId(optString(jsonObject, FIELD_PAYMENT_METHOD))
+                .setPaymentMethodTypes(jsonArrayToList(
+                        jsonObject.optJSONArray(FIELD_PAYMENT_METHOD_TYPES)))
+                .setStatus(Status.fromCode(optString(jsonObject, FIELD_STATUS)))
+                .setUsage(Usage.fromCode(optString(jsonObject, FIELD_USAGE)))
+                .setNextAction(optMap(jsonObject, FIELD_NEXT_ACTION))
+                .build();
+    }
+
+    @NonNull
+    @Override
+    public Map<String, Object> toMap() {
+        final AbstractMap<String, Object> map = new HashMap<>();
+        map.put(FIELD_ID, mId);
+        map.put(FIELD_OBJECT, mObjectType);
+        map.put(FIELD_PAYMENT_METHOD, mPaymentMethodId);
+        map.put(FIELD_PAYMENT_METHOD_TYPES, mPaymentMethodTypes);
+        map.put(FIELD_CLIENT_SECRET, mClientSecret);
+        map.put(FIELD_CREATED, mCreated);
+        map.put(FIELD_CUSTOMER, mCustomerId);
+        map.put(FIELD_DESCRIPTION, mDescription);
+        map.put(FIELD_LIVEMODE, mLiveMode);
+        map.put(FIELD_NEXT_ACTION, mNextAction);
+        map.put(FIELD_STATUS, mStatus != null ? mStatus.code : null);
+        map.put(FIELD_USAGE, mUsage != null ? mUsage.code : null);
+        StripeNetworkUtils.removeNullAndEmptyParams(map);
+        return map;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object obj) {
+        return this == obj || (obj instanceof SetupIntent && typedEquals((SetupIntent) obj));
+    }
+
+    private boolean typedEquals(@NonNull SetupIntent setupIntent) {
+        return ObjectUtils.equals(mId, setupIntent.mId)
+                && ObjectUtils.equals(mObjectType, setupIntent.mObjectType)
+                && ObjectUtils.equals(mClientSecret, setupIntent.mClientSecret)
+                && ObjectUtils.equals(mCreated, setupIntent.mCreated)
+                && ObjectUtils.equals(mCustomerId, setupIntent.mCustomerId)
+                && ObjectUtils.equals(mDescription, setupIntent.mDescription)
+                && ObjectUtils.equals(mLiveMode, setupIntent.mLiveMode)
+                && ObjectUtils.equals(mStatus, setupIntent.mStatus)
+                && ObjectUtils.equals(mUsage, setupIntent.mUsage)
+                && ObjectUtils.equals(mPaymentMethodId, setupIntent.mPaymentMethodId)
+                && ObjectUtils.equals(mPaymentMethodTypes, setupIntent.mPaymentMethodTypes)
+                && ObjectUtils.equals(mNextAction, setupIntent.mNextAction)
+                && ObjectUtils.equals(mNextActionType, setupIntent.mNextActionType);
+    }
+
+    @Override
+    public int hashCode() {
+        return ObjectUtils.hash(mId, mObjectType, mCustomerId, mClientSecret, mCreated,
+                mDescription, mLiveMode, mStatus, mPaymentMethodId, mPaymentMethodTypes,
+                mNextAction, mNextActionType, mUsage);
+    }
+
+    static final class Builder implements ObjectBuilder<SetupIntent> {
```

Comment:
nice! this looks great :)

i think the underlying `ObjectBuilder` could probably use a couple extra methods like `build` and `isValid` or something so that we don't have to have `mPaymentMethodTypes` in the Builder be nullable but `getPaymentMethodTypes` is `NonNull` for example - but that's beyond the scope of this patch.

thanks for doing this!